File: Java/003_Longest Substring Without Repeating Characters.java
package algorithm;

import java.util.*;

class ListNode {
	int val;
	ListNode next;
	ListNode(int x) {
		val = x;
		next = null;
	}
}

class Solution {
	public int lengthOfLongestSubstring(String s) {
		HashMap<Character, Byte> hm = new HashMap<Character, Byte>();
		hm.clear();
		int i = 0, j = 0;
		int res = 0;
		while (j < s.length()) {
			if (hm.get(s.charAt(j)) == null) {
				hm.put(s.charAt(j), (byte)0);
				j++;
				res = Math.max(res, j - i);
			}
			else {
				hm.remove(s.charAt(i));
				i++;
			}
		}
		return res;
	}
}

public class Test {
	public static void main(String[] args) {
		Solution s = new Solution();
		System.out.println(s.lengthOfLongestSubstring("abcabcbb"));
		System.out.println(s.lengthOfLongestSubstring("bbbbb"));
	}
}

File: Python/003_Longest Substring Without Repeating Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        appear = dict()
        i, j, l = 0, 0, len(s)
        ma, num = 0, 0
        while j < l:
            if not appear.get(s[j], 0):
                appear[s[j]] = 1
            else:
                while s[i] != s[j]:
                    appear.pop(s[i], None)
                    i, num = i + 1, num - 1
                i, num = i + 1, num - 1
            num, j = num + 1, j + 1
            ma = max(ma, num)
        return ma